File: src/web/hospital-web/src/app/core/services/search.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import { Observable } from 'rxjs';

export interface DoctorSearch {
  id: number;
  name: string;
  specializationId: number;
  experienceYears: number;
  contact: string;
}

export interface PatientSearch {
  id: number;
  name: string;
  condition: string;
  gender: string;
  contact: string;
}

export interface RecordSearch {
  id: number;
  patientId: number;
  doctorId: number;
  recordType: string;
  isArchived: boolean;
  createdOn: string;
}

@Injectable({ providedIn: 'root' })
export class SearchService {
  private base = `${environment.apiBaseUrl}/search`;

  constructor(private http: HttpClient) {}

  searchDoctors(name?: string, specializationId?: number): Observable<DoctorSearch[]> {
    let params = new HttpParams();
    if (name) params = params.set('name', name);
    if (specializationId != null) params = params.set('specializationId', specializationId);
    return this.http.get<DoctorSearch[]>(`${this.base}/doctors`, { params });
  }

  searchPatients(name?: string, condition?: string): Observable<PatientSearch[]> {
    let params = new HttpParams();
    if (name) params = params.set('name', name);
    if (condition) params = params.set('condition', condition);
    return this.http.get<PatientSearch[]>(`${this.base}/patients`, { params });
  }

  searchRecords(patientId?: number, doctorId?: number, includeArchived = false): Observable<RecordSearch[]> {
    let params = new HttpParams().set('includeArchived', includeArchived);
    if (patientId != null) params = params.set('patientId', patientId);
    if (doctorId != null) params = params.set('doctorId', doctorId);
    return this.http.get<RecordSearch[]>(`${this.base}/records`, { params });
  }
}

File: src/Hospital.Infrastructure/Services/SearchService.cs
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using System.Collections.Generic;
using Hospital.Application.DTO.Search;
using Hospital.Application.Services;
using Hospital.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;

namespace Hospital.Infrastructure.Services
{
    public class SearchService : ISearchService
    {
        private readonly ApplicationDbContext _db;

        public SearchService(ApplicationDbContext db)
        {
            _db = db;
        }

        public async Task<List<DoctorSearchDto>> SearchDoctors(string? name, int? specializationId, CancellationToken ct)
        {
            var q = _db.Doctors
                .AsNoTracking()
                .Where(x => !x.IsDeleted);

            if (!string.IsNullOrWhiteSpace(name))
                q = q.Where(x => x.Name.Contains(name));

            if (specializationId.HasValue)
                q = q.Where(x => x.SpecializationId == specializationId.Value);

            return await q
                .OrderBy(x => x.Name)
                .Select(x => new DoctorSearchDto
                {
                    Id = x.Id,
                    Name = x.Name,
                    SpecializationId = x.SpecializationId,
                    ExperienceYears = x.ExperienceYears,
                    Contact = x.Contact
                })
                .ToListAsync(ct);
        }

        public async Task<List<PatientSearchDto>> SearchPatients(string? name, string? condition, CancellationToken ct)
        {
            var q = _db.Patients
                .AsNoTracking()
                .Where(x => !x.IsDeleted);

            if (!string.IsNullOrWhiteSpace(name))
                q = q.Where(x => x.Name.Contains(name));

            if (!string.IsNullOrWhiteSpace(condition))
                q = q.Where(x => x.Condition.Contains(condition));

            return await q
                .OrderBy(x => x.Name)
                .Select(x => new PatientSearchDto
                {
                    Id = x.Id,
                    Name = x.Name,
                    Condition = x.Condition,
                    Gender = x.Gender,
                    Contact = x.Contact
                })
                .ToListAsync(ct);
        }

        public async Task<List<RecordSearchDto>> SearchRecords(int? patientId, int? doctorId, bool includeArchived, CancellationToken ct)
        {
            var q = _db.MedicalRecords
                .AsNoTracking()
                .Where(x => !x.IsDeleted);

            if (!includeArchived)
                q = q.Where(x => !x.IsArchived);

            if (patientId.HasValue)
                q = q.Where(x => x.PatientId == patientId.Value);

            if (doctorId.HasValue)
                q = q.Where(x => x.DoctorId == doctorId.Value);

            return await q
                .OrderByDescending(x => x.CreatedOn)
                .Select(x => new RecordSearchDto
                {
                    Id = x.Id,
                    PatientId = x.PatientId,
                    DoctorId = x.DoctorId,
                    RecordType = x.RecordType,
                    IsArchived = x.IsArchived,
                    CreatedOn = x.CreatedOn
                })
                .ToListAsync(ct);
        }
    }
}
